src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,110 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final double GROW_COEFFICIENT = 0.75;
+    private static final byte DEFAULT_SIZE = 10;
+    private int size = DEFAULT_SIZE;
+    private int currentIndex = 0;
+    private T[] storage;
+
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        storage = (T[]) new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        shouldGrow();
+        storage[currentIndex++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > currentIndex || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    String.format("Index: %d, Size: %d", index, currentIndex)
+            );
+        }
+        shouldGrow();
 
+        System.arraycopy(storage, index, storage, index + 1, currentIndex - index);
+        storage[index] = value;
+        currentIndex++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int j = 0; j < list.size(); j++) {
+            add(list.get(j));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return storage[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        storage[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T removedElement = storage[index];
```

Comment:
You should cast the value to T

Reply:
It's redundant code casting, because we can't pass not T type values in storage. What do you think ?

Reply:
Yes, you are right! I overlooked that you had an array of T-type values, but not of Objects